pkg/container-collection/container-collection.go

```diff
@@ -0,0 +1,228 @@
+// Copyright 2019-2021 The Inspektor Gadget authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// Package containercollection provides the ContainerCollection struct to keep
+// track of the set of running containers and primitives to query that set with
+// various criteria.
+//
+// It is used by the Gadget Tracer Manager to keep track of containers part of
+// Kubernetes pods and by Local Gadget Manager to keep track of containers on a
+// Linux system.
+package containercollection
+
+import (
+	"sync"
+
+	pb "github.com/kinvolk/inspektor-gadget/pkg/gadgettracermanager/api"
+	"github.com/kinvolk/inspektor-gadget/pkg/gadgettracermanager/pubsub"
+)
+
+// ContainerCollection holds a set of containers. It can be embedded as an
+// anonymous struct to help other structs implement the ContainerResolver
+// interface. For this reason, some methods are prefixed with 'Container' to
+// make clear that
+
+type ContainerCollection struct {
+	// Keys:   Container ID
+	// Values: pb.ContainerDefinition
+	containers sync.Map
```

Comment:
Do we really need to use this kind of map? Do we have a big lock contention? Perhaps using a mutex + standard map will be more readable and the performance difference insignificant?

Reply:
Before this PR, the Gadget Tracer Manager was using a mutex + 2 standard maps for `g.containers` and `g.tracers`. But then, some functions were iterating over all containers with the mutex taken and calling a callback external to the package while the mutex was taken (see [Subscribe](https://github.com/kinvolk/inspektor-gadget/blob/f5dcc6aaa6e66debfe7ed93408c66ed1b003c1ec/pkg/gadgettracermanager/gadgettracermanager.go#L579-L584)). Then, the callbacks were re-entering into the Gadget Tracer Manager code and trying to take the lock again, leading to a deadlock. So I disabled the lock for now, leading to incorrect code (See [PublishEvent](https://github.com/kinvolk/inspektor-gadget/blob/f5dcc6aaa6e66debfe7ed93408c66ed1b003c1ec/pkg/gadgettracermanager/gadgettracermanager.go#L257-L259)).

Example: dns gadget:
1. Call [Subscribe](https://github.com/kinvolk/inspektor-gadget/blob/f5dcc6aaa6e66debfe7ed93408c66ed1b003c1ec/pkg/gadgets/dns/gadget.go#L213-L217)
2. Call [containerEventCallback](https://github.com/kinvolk/inspektor-gadget/blob/f5dcc6aaa6e66debfe7ed93408c66ed1b003c1ec/pkg/gadgets/dns/gadget.go#L204)
3. Call [attachContainerFunc](https://github.com/kinvolk/inspektor-gadget/blob/f5dcc6aaa6e66debfe7ed93408c66ed1b003c1ec/pkg/gadgets/dns/gadget.go#L172)
4. Call [PublishEvent](https://github.com/kinvolk/inspektor-gadget/blob/f5dcc6aaa6e66debfe7ed93408c66ed1b003c1ec/pkg/gadgets/dns/gadget.go#L181)

The seccomp gadget has something similar but not merged yet.

I think `sync.Map` solves this kind of situations because it provides the [Range()](https://pkg.go.dev/sync#Map.Range) method to iterate over the elements and allowing to call a callback without the lock taken. The standard map does not allow that.

Once we do something similar for the `g.tracers`, then we can fix the issue left as a TODO

Reply:
That's a good point I wasn't taking into consideration. It makes sense to continue using sync.Map then.